File: app/src/main/java/com/zack/enderplan/domain/activity/TypeDetailActivity.java
package com.zack.enderplan.domain.activity;

import android.animation.ObjectAnimator;
import android.app.ActivityOptions;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.zack.enderplan.R;
import com.zack.enderplan.domain.view.TypeDetailView;
import com.zack.enderplan.interactor.adapter.SingleTypePlanAdapter;
import com.zack.enderplan.interactor.adapter.SimpleTypeAdapter;
import com.zack.enderplan.interactor.callback.PlanItemTouchCallback;
import com.zack.enderplan.interactor.presenter.TypeDetailPresenter;
import com.zack.enderplan.model.bean.FormattedType;
import com.zack.enderplan.model.bean.Plan;
import com.zack.enderplan.utility.Util;
import com.zack.enderplan.widget.CircleColorView;

import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class TypeDetailActivity extends BaseActivity implements TypeDetailView {

    @BindView(R.id.layout_app_bar)
    AppBarLayout mAppBarLayout;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.layout_header)
    LinearLayout mHeaderLayout;
    @BindView(R.id.ic_type_mark)
    CircleColorView typeMarkIcon;
    @BindView(R.id.text_type_name)
    TextView typeNameText;
    @BindView(R.id.text_uc_plan_count)
    TextView ucPlanCountText;
    @BindView(R.id.editor_content)
    EditText contentEditor;
    @BindView(R.id.list_plan)
    RecyclerView mPlanList;
    @BindView(R.id.layout_editor)
    FrameLayout mEditorLayout;

    @BindString(R.string.hint_editor_content_format)
    String mContentEditorHintFormat;
    @BindString(R.string.snackbar_delete_format)
    String mSnackbarDeleteFormat;
    @BindString(R.string.name_type_mark_shared_element_transition)
    String mTypeMarkSetName;

    private TypeDetailPresenter typeDetailPresenter;
    
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        typeDetailPresenter = new TypeDetailPresenter(this, getIntent().getIntExtra("position", -1));
        typeDetailPresenter.setInitialView();
    }

    @Override
    protected void onResume() {
        super.onResume();
        typeDetailPresenter.notifySwitchingViewVisibility(true);
    }

    @Override
    protected void onPause() {
        super.onPause();
        typeDetailPresenter.notifySwitchingViewVisibility(false);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        typeDetailPresenter.detachView();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_type_detail, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            case R.id.action_edit:
                typeDetailPresenter.notifyTypeEditingButtonClicked();
                break;
            case R.id.action_delete:
                //TODO 可能需要收起软键盘
                typeDetailPresenter.notifyTypeDeletionButtonClicked(false);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        typeDetailPresenter.notifyBackPressed();
    }

    @Override
    public void showInitialView(FormattedType formattedType, String ucPlanCountStr, SingleTypePlanAdapter singleTypePlanAdapter) {
        setContentView(R.layout.activity_type_detail);
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        setupActionBar();

        mAppBarLayout.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                mAppBarLayout.getViewTreeObserver().removeOnPreDrawListener(this);
                typeDetailPresenter.notifyPreDrawingAppBar(mAppBarLayout.getTotalScrollRange());
                return false;
            }
        });

        mAppBarLayout.addOnOffsetChangedListener(new AppBarLayout.OnOffsetChangedListener() {
            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                typeDetailPresenter.notifyAppBarScrolled(verticalOffset);
            }
        });

        mEditorLayout.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                mEditorLayout.getViewTreeObserver().removeOnPreDrawListener(this);
                typeDetailPresenter.notifyPreDrawingEditorLayout(mEditorLayout.getHeight());
                return false;
            }
        });

        singleTypePlanAdapter.setOnPlanItemClickListener(new SingleTypePlanAdapter.OnPlanItemClickListener() {
            @Override
            public void onPlanItemClick(int position) {
                typeDetailPresenter.notifyPlanItemClicked(position);
            }
        });

        singleTypePlanAdapter.setOnStarButtonClickListener(new SingleTypePlanAdapter.OnStarButtonClickListener() {
            @Override
            public void onStarButtonClick(int position) {
                typeDetailPresenter.notifyPlanStarStatusChanged(position);
            }
        });

        mPlanList.setLayoutManager(new LinearLayoutManager(this));
        mPlanList.setHasFixedSize(true);
        mPlanList.setAdapter(singleTypePlanAdapter);

        PlanItemTouchCallback planItemTouchCallback = new PlanItemTouchCallback();
        planItemTouchCallback.setOnItemSwipedListener(new PlanItemTouchCallback.OnItemSwipedListener() {
            @Override
            public void onItemSwiped(int position, int direction) {
                switch (direction) {
                    case PlanItemTouchCallback.DIR_START:
                        typeDetailPresenter.notifyDeletingPlan(position);
                        break;
                    case PlanItemTouchCallback.DIR_END:
                        typeDetailPresenter.notifySwitchingPlanStatus(position);
                        break;
                }
            }
        });
        planItemTouchCallback.setOnItemMovedListener(new PlanItemTouchCallback.OnItemMovedListener() {
            @Override
            public void onItemMoved(int fromPosition, int toPosition) {
                typeDetailPresenter.notifyPlanSequenceChanged(fromPosition, toPosition);
            }
        });
        new ItemTouchHelper(planItemTouchCallback).attachToRecyclerView(mPlanList);

        contentEditor.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    typeDetailPresenter.notifyCreatingPlan(v.getText().toString());
                }
                return false;
            }
        });

        onTypeNameChanged(formattedType.getTypeName(), formattedType.getFirstChar());
        onTypeMarkColorChanged(formattedType.getTypeMarkColorInt());
        onTypeMarkPatternChanged(formattedType.isHasTypeMarkPattern(), formattedType.getTypeMarkPatternResId());
        onUcPlanCountChanged(ucPlanCountStr);
    }

    @Override
    public void onTypeNameChanged(String typeName, String firstChar) {
        typeMarkIcon.setInnerText(firstChar);
        typeNameText.setText(typeName);
        contentEditor.setHint(String.format(mContentEditorHintFormat, typeName));
    }

    @Override
    public void onTypeMarkColorChanged(int colorInt) {
        typeMarkIcon.setFillColor(colorInt);
    }

    @Override
    public void onTypeMarkPatternChanged(boolean hasPattern, int patternResId) {
        typeMarkIcon.setInnerIcon(hasPattern ? getDrawable(patternResId) : null);
    }

    @Override
    public void onPlanCreated() {
        mPlanList.scrollToPosition(0);
        contentEditor.setText("");
    }

    @Override
    public void onUcPlanCountChanged(String ucPlanCountStr) {
        ucPlanCountText.setText(ucPlanCountStr);
    }

    @Override
    public void onPlanDeleted(final Plan deletedPlan, final int position, final int planListPos, boolean shouldShowSnackbar) {
        if (shouldShowSnackbar) {
            Snackbar.make(mPlanList, String.format(mSnackbarDeleteFormat, deletedPlan.getContent()), Snackbar.LENGTH_LONG)
                    .setAction(R.string.button_undo, new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            typeDetailPresenter.notifyCreatingPlan(deletedPlan, position, planListPos);
                        }
                    })
                    .show();
        }
    }

    @Override
    public void onPlanItemClicked(int posInPlanList) {
        Intent intent = new Intent(this, PlanDetailActivity.class);
        intent.putExtra("position", posInPlanList);
        ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(this);
        startActivity(intent, options.toBundle());
    }

    @Override
    public void onAppBarScrolled(float headerLayoutAlpha) {
        mHeaderLayout.setAlpha(headerLayoutAlpha);
    }

    @Override
    public void onAppBarScrolledToCriticalPoint(String toolbarTitle, float editorLayoutTransY) {
        toolbar.setTitle(toolbarTitle);
        ObjectAnimator.ofFloat(mEditorLayout, "translationY", mEditorLayout.getTranslationY(), editorLayoutTransY)
                .setDuration(200)
                .start();
    }

    @Override
    public void backToTop() {
        mAppBarLayout.setExpanded(true);
        //TODO 滑动到顶部
        //mPlanList.scrollToPosition(0);
    }

    @Override
    public void pressBack() {
        super.onBackPressed();
    }

    @Override
    public void enterEditType(int position, boolean shouldPlaySharedElementTransition) {
        Intent intent = new Intent(this, EditTypeActivity.class);
        intent.putExtra("position", position);
        ActivityOptions options;
        if (shouldPlaySharedElementTransition) {
            options = ActivityOptions.makeSceneTransitionAnimation(this, typeMarkIcon, mTypeMarkSetName);
        } else {
            options = ActivityOptions.makeSceneTransitionAnimation(this);
        }
        startActivity(intent, options.toBundle());
    }

    @Override
    public void showToast(int msgResId) {
        Toast.makeText(this, msgResId, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onDetectedDeletingLastType() {
        new AlertDialog.Builder(this)
                .setTitle(R.string.title_dialog_last_type)
                .setMessage(R.string.msg_dialog_last_type)
                .setPositiveButton(R.string.button_ok, null)
                .show();
    }

    @Override
    public void onDetectedTypeNotEmpty() {
        new AlertDialog.Builder(this)
                .setTitle(R.string.title_dialog_type_not_empty)
                .setMessage(Util.addBoldStyle(getString(R.string.msg_dialog_type_not_empty), new String[]{getString(R.string.button_move), getString(R.string.button_delete), getString(R.string.button_cancel)}))
                .setPositiveButton(R.string.button_move, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        typeDetailPresenter.notifyMovePlanButtonClicked();
                    }
                })
                .setNeutralButton(R.string.button_delete, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        typeDetailPresenter.notifyTypeDeletionButtonClicked(true);
                    }
                })
                .setNegativeButton(R.string.button_cancel, null)
                .show();
    }

    @Override
    public void showMovePlanDialog(int planCount, SimpleTypeAdapter simpleTypeAdapter) {
        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.title_dialog_move_plan_pt1) + " " + planCount + " " + getString(planCount > 1 ? R.string.title_dialog_move_plan_pt2_pl : R.string.title_dialog_move_plan_pt2_sg))
                .setAdapter(simpleTypeAdapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        typeDetailPresenter.notifyTypeItemInMovePlanDialogClicked(which);
                    }
                })
                .show();
    }

    @Override
    public void showTypeDeletionConfirmationDialog(String typeName) {
        new AlertDialog.Builder(this)
                .setTitle(typeName)
                .setMessage(R.string.msg_dialog_delete_type)
                .setPositiveButton(R.string.button_delete, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        typeDetailPresenter.notifyDeletingType(false, null);
                    }
                })
                .setNegativeButton(R.string.button_cancel, null)
                .show();
    }

    @Override
    public void showPlanMigrationConfirmationDialog(String fromTypeName, String toTypeName, final String toTypeCode) {
        new AlertDialog.Builder(this)
                .setTitle(fromTypeName)
                .setMessage(Util.addBoldStyle(getString(R.string.msg_dialog_migrate_plan_pt1) + " " + toTypeName + getString(R.string.msg_dialog_migrate_plan_pt2), new String[]{toTypeName}))
                .setPositiveButton(R.string.btn_dialog_move_and_delete, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        typeDetailPresenter.notifyDeletingType(true, toTypeCode);
                    }
                })
                .setNegativeButton(R.string.button_cancel, null)
                .show();
    }

    @Override
    public void exitTypeDetail() {
        finish();
    }

    @OnClick({R.id.ic_clear_text})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.ic_clear_text:
                contentEditor.setText("");
                break;
        }
    }
}

File: app/src/main/java/net/zackzhang/app/end/view/activity/TypeDetailActivity.kt
package net.zackzhang.app.end.view.activity

import android.animation.ObjectAnimator
import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.view.inputmethod.EditorInfo
import butterknife.ButterKnife
import butterknife.OnClick
import kotlinx.android.synthetic.main.activity_type_detail.*
import kotlinx.android.synthetic.main.content_type_detail.*
import net.zackzhang.app.end.App
import net.zackzhang.app.end.R
import net.zackzhang.app.end.common.Constant
import net.zackzhang.app.end.injector.component.DaggerTypeDetailComponent
import net.zackzhang.app.end.injector.module.TypeDetailPresenterModule
import net.zackzhang.app.end.model.bean.FormattedType
import net.zackzhang.app.end.model.bean.Plan
import net.zackzhang.app.end.presenter.TypeDetailPresenter
import net.zackzhang.app.end.util.ColorUtil
import net.zackzhang.app.end.util.ResourceUtil
import net.zackzhang.app.end.util.StringUtil
import net.zackzhang.app.end.view.adapter.SingleTypePlanListAdapter
import net.zackzhang.app.end.view.contract.TypeDetailViewContract
import net.zackzhang.app.end.view.dialog.MessageDialogFragment
import net.zackzhang.app.end.view.dialog.TypePickerForPlanMigrationDialogFragment
import javax.inject.Inject

class TypeDetailActivity : BaseActivity(), TypeDetailViewContract {

    companion object {

        private const val TAG_DETECTED_TYPE_NOT_EMPTY = "detected_type_not_empty"
        private const val TAG_MOVE_PLAN = "move_plan"
        private const val TAG_TYPE_DELETION_CONFIRMATION = "type_deletion_confirmation"
        private const val TAG_PLAN_MIGRATION_CONFIRMATION = "type_migration_confirmation"

        private const val KEY_PLAN_MIGRATION_TO_TYPE_CODE = "plan_migration_to_type_code"

        fun start(activity: Activity, typeListPosition: Int, enableTransition: Boolean, sharedElement: View?, transitionName: String?) {
            val intent = Intent(activity, TypeDetailActivity::class.java)
            intent.putExtra(Constant.TYPE_LIST_POSITION, typeListPosition)
            intent.putExtra(Constant.ENABLE_TRANSITION, enableTransition)
            intent.putExtra(Constant.TRANSITION_NAME, transitionName)
            if (enableTransition) {
                activity.startActivity(
                        intent,
                        ActivityOptions.makeSceneTransitionAnimation(activity, sharedElement, transitionName).toBundle()
                )
            } else {
                activity.startActivity(intent)
            }
        }
    }

    @Inject
    lateinit var mTypeDetailPresenter: TypeDetailPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mTypeDetailPresenter.attach()
    }

    override fun onInjectPresenter() {
        DaggerTypeDetailComponent.builder()
                .typeDetailPresenterModule(TypeDetailPresenterModule(this, intent.getIntExtra(Constant.TYPE_LIST_POSITION, -1)))
                .appComponent(App.appComponent)
                .build()
                .inject(this)
    }

    override fun onResume() {
        super.onResume()
        mTypeDetailPresenter.notifySwitchingViewVisibility(true)
    }

    override fun onPause() {
        super.onPause()
        mTypeDetailPresenter.notifySwitchingViewVisibility(false)
    }

    override fun onDestroy() {
        super.onDestroy()
        mTypeDetailPresenter.detach()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_type_detail, menu)
        //在这里改变图标的tint，因为没法在xml文件中改
        menu.findItem(R.id.action_edit).icon.setTint(Color.WHITE)
        menu.findItem(R.id.action_delete).icon.setTint(Color.WHITE)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.action_edit -> mTypeDetailPresenter.notifyTypeEditingButtonClicked()
            //TODO 可能需要收起软键盘
            R.id.action_delete -> mTypeDetailPresenter.notifyTypeDeletionButtonClicked(false)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        mTypeDetailPresenter.notifyBackPressed()
    }

    override fun onAttachFragment(fragment: Fragment) {
        super.onAttachFragment(fragment)

        when (fragment.tag) {
            TAG_DETECTED_TYPE_NOT_EMPTY -> {
                val dialogFragment = fragment as MessageDialogFragment
                dialogFragment.okButtonClickListener = { mTypeDetailPresenter.notifyMovePlanButtonClicked() }
                dialogFragment.thirdButtonClickListener = { mTypeDetailPresenter.notifyTypeDeletionButtonClicked(true) }
            }
            TAG_MOVE_PLAN -> (fragment as TypePickerForPlanMigrationDialogFragment).typePickedListener = { typeCode, typeName -> mTypeDetailPresenter.notifyTypeItemInMovePlanDialogClicked(typeCode, typeName) }
            TAG_TYPE_DELETION_CONFIRMATION -> (fragment as MessageDialogFragment).okButtonClickListener = { mTypeDetailPresenter.notifyDeletingType(false, null) }
            TAG_PLAN_MIGRATION_CONFIRMATION -> {
                val dialogFragment = fragment as MessageDialogFragment
                dialogFragment.okButtonClickListener = { mTypeDetailPresenter.notifyDeletingType(true, dialogFragment.arguments!!.getString(KEY_PLAN_MIGRATION_TO_TYPE_CODE)) }
            }
        }
    }

    override fun showInitialView(formattedType: FormattedType, ucPlanCountStr: String, singleTypePlanListAdapter: SingleTypePlanListAdapter, itemTouchHelper: ItemTouchHelper) {
        val enableTransition = intent.getBooleanExtra(Constant.ENABLE_TRANSITION, false)

        if (enableTransition) {
            window.requestFeature(Window.FEATURE_CONTENT_TRANSITIONS)
        }

        setContentView(R.layout.activity_type_detail)
        ButterKnife.bind(this)

        setSupportActionBar(toolbar)
        setupActionBar()

        layout_app_bar.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                layout_app_bar.viewTreeObserver.removeOnPreDrawListener(this)
                mTypeDetailPresenter.notifyPreDrawingAppBar(layout_app_bar.totalScrollRange)
                return false
            }
        })

        layout_app_bar.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { _, verticalOffset -> mTypeDetailPresenter.notifyAppBarScrolled(verticalOffset) })

        if (enableTransition) {
            ic_type_mark.transitionName = intent.getStringExtra(Constant.TRANSITION_NAME)
        }

        layout_editor.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                layout_editor.viewTreeObserver.removeOnPreDrawListener(this)
                mTypeDetailPresenter.notifyPreDrawingEditorLayout(layout_editor.height)
                return false
            }
        })

        list_single_type_plan.adapter = singleTypePlanListAdapter
        list_single_type_plan.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                mTypeDetailPresenter.notifyPlanListScrolled(
                        !list_single_type_plan.canScrollVertically(-1),
                        !list_single_type_plan.canScrollVertically(1)
                )
            }
        })
        itemTouchHelper.attachToRecyclerView(list_single_type_plan)

        editor_content.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable) {
                mTypeDetailPresenter.notifyContentEditorTextChanged(s.toString())
            }
        })
        editor_content.setOnEditorActionListener { v, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                mTypeDetailPresenter.notifyCreatingPlan(v.text.toString())
            }
            false
        }

        onTypeNameChanged(formattedType.typeName, formattedType.firstChar)
        onTypeMarkColorChanged(formattedType.typeMarkColorInt)
        onTypeMarkPatternChanged(formattedType.hasTypeMarkPattern, formattedType.typeMarkPatternResId)
        onUcPlanCountChanged(ucPlanCountStr)
    }

    override fun onTypeNameChanged(typeName: String, firstChar: String) {
        ic_type_mark.setInnerText(firstChar)
        text_type_name.text = typeName
        editor_content.hint = String.format(getString(R.string.hint_editor_content_format), typeName)
    }

    override fun onTypeMarkColorChanged(colorInt: Int) {
        window.navigationBarColor = colorInt
        val headerColorInt = ColorUtil.reduceSaturation(colorInt, 0.85f)
        layout_collapsing_toolbar.setContentScrimColor(headerColorInt)
        layout_collapsing_toolbar.setStatusBarScrimColor(colorInt)
        bg_header.setImageDrawable(ColorDrawable(headerColorInt))
        ic_type_mark.setFillColor(colorInt)
    }

    override fun onTypeMarkPatternChanged(hasPattern: Boolean, patternResId: Int) {
        ic_type_mark.setInnerIcon(if (hasPattern) getDrawable(patternResId) else null)
    }

    override fun onPlanCreated() {
        list_single_type_plan.scrollToPosition(0)
        editor_content.text = null
    }

    override fun onUcPlanCountChanged(ucPlanCountStr: String) {
        text_uc_plan_count.text = ucPlanCountStr
    }

    override fun onPlanDeleted(deletedPlan: Plan, position: Int, planListPos: Int, shouldShowSnackbar: Boolean) {
        if (shouldShowSnackbar) {
            Snackbar.make(list_single_type_plan, String.format(getString(R.string.snackbar_delete_format), deletedPlan.content), Snackbar.LENGTH_LONG)
                    .setAction(R.string.button_undo) { mTypeDetailPresenter.notifyCreatingPlan(deletedPlan, position, planListPos) }
                    .show()
        }
    }

    override fun onPlanItemClicked(posInPlanList: Int) {
        PlanDetailActivity.start(this, posInPlanList)
    }

    override fun onAppBarScrolled(headerLayoutAlpha: Float) {
        layout_header.alpha = headerLayoutAlpha
    }

    override fun onAppBarScrolledToCriticalPoint(toolbarTitle: String, editorLayoutTransY: Float) {
        toolbar.title = toolbarTitle
        ObjectAnimator.ofFloat(layout_editor, "translationY", layout_editor.translationY, editorLayoutTransY)
                .setDuration(200)
                .start()
    }

    override fun changeContentEditorClearTextIconVisibility(isVisible: Boolean) {
        ic_clear_text.visibility = if (isVisible) View.VISIBLE else View.GONE
    }

    override fun backToTop() {
        layout_app_bar.setExpanded(true)
        //TODO 滑动到顶部
        //list_single_type_plan.scrollToPosition(0);
    }

    override fun pressBack() {
        super.onBackPressed()
    }

    override fun enterEditType(position: Int, enableTransition: Boolean) {
        TypeEditActivity.start(
                this,
                position,
                enableTransition,
                ic_type_mark,
                getString(R.string.transition_type_mark_icon)
        )
    }

    override fun onDetectedDeletingLastType() {
        MessageDialogFragment.Builder()
                .setTitle(R.string.title_dialog_last_type)
                .setMessage(R.string.msg_dialog_last_type)
                .setOkButtonText(android.R.string.ok)
                .show(supportFragmentManager)
    }

    override fun onDetectedTypeNotEmpty(planCount: Int) {
        val buttons = arrayOf(getString(R.string.button_move), getString(R.string.button_delete), getString(R.string.button_cancel))
        MessageDialogFragment.Builder()
                .setTitle(R.string.title_dialog_type_not_empty)
                .setMessage(StringUtil.addSpan(
                        StringUtil.toUpperCase(ResourceUtil.getQuantityString(R.string.msg_dialog_type_not_empty, R.plurals.text_plan_count, planCount), buttons),
                        buttons,
                        StringUtil.SPAN_BOLD_STYLE
                ))
                .setThirdButtonText(buttons[1])
                .setOkButtonText(buttons[0])
                .show(supportFragmentManager, TAG_DETECTED_TYPE_NOT_EMPTY)
    }

    override fun showMovePlanDialog(typeCode: String) {
        TypePickerForPlanMigrationDialogFragment.newInstance(typeCode).show(supportFragmentManager, TAG_MOVE_PLAN)
    }

    override fun showTypeDeletionConfirmationDialog(typeName: String) {
        MessageDialogFragment.Builder()
                .setTitle(typeName)
                .setMessage(R.string.msg_dialog_delete_type)
                .setOkButtonText(R.string.button_delete)
                .showCancelButton()
                .show(supportFragmentManager, TAG_TYPE_DELETION_CONFIRMATION)
    }

    override fun showPlanMigrationConfirmationDialog(fromTypeName: String, planCount: Int, toTypeName: String, toTypeCode: String) {
        val dialogFragment = MessageDialogFragment.Builder()
                .setTitle(fromTypeName)
                .setMessage(StringUtil.addSpan(
                        String.format(ResourceUtil.getString(R.string.msg_dialog_migrate_plan), ResourceUtil.getQuantityString(R.plurals.text_plan_count, planCount), toTypeName, fromTypeName),
                        arrayOf(toTypeName, fromTypeName),
                        StringUtil.SPAN_BOLD_STYLE
                ))
                .setOkButtonText(R.string.btn_dialog_move_and_delete)
                .showCancelButton()
                .build()
        // 将 toTypeCode 直接存入 dialogFragment 的 arguments 中，供确定按钮的回调函数使用
        dialogFragment.arguments!!.putString(KEY_PLAN_MIGRATION_TO_TYPE_CODE, toTypeCode)
        dialogFragment.show(supportFragmentManager, TAG_PLAN_MIGRATION_CONFIRMATION)
    }

    @OnClick(R.id.ic_clear_text)
    fun onClick(view: View) {
        when (view.id) {
            R.id.ic_clear_text -> editor_content.setText("")
        }
    }
}
